File: backend/src/main/java/com/samarthshastry/workspace/backend/service/WorkspaceService.java
package com.samarthshastry.workspace.backend.service;

import com.samarthshastry.workspace.backend.model.Workspace;
import com.samarthshastry.workspace.backend.model.User;
import com.samarthshastry.workspace.backend.repository.WorkspaceRepo;
import com.samarthshastry.workspace.backend.repository.UserRepo;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class WorkspaceService {

    private final WorkspaceRepo workspaceRepository;
    private final UserRepo userRepository;

    public List<Workspace> getAllWorkspaces() {
        return workspaceRepository.findAll();
    }

    public Optional<Workspace> getWorkspaceById(UUID id) {
        return workspaceRepository.findById(id);
    }

    public List<Workspace> getWorkspacesByOwner(UUID ownerId) {
        User owner = userRepository.findById(ownerId)
                .orElseThrow(() -> new RuntimeException("Owner not found"));
        return workspaceRepository.findByOwner(owner);
    }

    public List<Workspace> getWorkspacesByMember(UUID userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        return workspaceRepository.findByMembersContaining(user);
    }

    @Transactional
    public Workspace createWorkspace(String name, String description, UUID ownerId) {
        User owner = userRepository.findById(ownerId)
                .orElseThrow(() -> new RuntimeException("Owner not found"));

        if (workspaceRepository.existsByNameAndOwner(name, owner)) {
            throw new RuntimeException("Workspace with this name already exists for this owner");
        }

        Workspace workspace = new Workspace();
        workspace.setName(name);
        workspace.setDescription(description);
        workspace.setOwner(owner);
        
        workspace.addMember(owner);

        return workspaceRepository.save(workspace);
    }

    @Transactional
    public Workspace addMember(UUID workspaceId, UUID userId) {
        Workspace workspace = workspaceRepository.findById(workspaceId)
                .orElseThrow(() -> new RuntimeException("Workspace not found"));
        
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        workspace.addMember(user);
        return workspaceRepository.save(workspace);
    }

    @Transactional
    public Workspace removeMember(UUID workspaceId, UUID userId) {
        Workspace workspace = workspaceRepository.findById(workspaceId)
                .orElseThrow(() -> new RuntimeException("Workspace not found"));
        
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        // Prevent removing the owner
        if (workspace.getOwner().getId().equals(userId)) {
            throw new RuntimeException("Cannot remove the owner from the workspace");
        }

        workspace.removeMember(user);
        return workspaceRepository.save(workspace);
    }

    @Transactional
    public Workspace updateWorkspaceName(UUID workspaceID, String name) {
        Workspace workspace = workspaceRepository.findById(workspaceID)
                .orElseThrow(() -> new RuntimeException("Workspace not found"));

        if (!name.equals(workspace.getName())) {
            workspace.setName(name);
            return workspaceRepository.save(workspace);
        } else {
            return workspace;
        }
    }

    @Transactional
    public Workspace updateWorkspaceDescription(UUID workspaceID, String description) {
        Workspace workspace = workspaceRepository.findById(workspaceID)
                .orElseThrow(() -> new RuntimeException("Workspace not found"));

        if (!description.equals(workspace.getDescription())) {
            workspace.setDescription(description);
            return workspaceRepository.save(workspace);
        } else {
            return workspace;
        }
    }

    public void deleteWorkspace(UUID id) {
        workspaceRepository.deleteById(id);
    }
}

File: frontend/src/app/core/services/workspace.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Workspace {
  id: string;
  name: string;
  description: string;
  owner: {
    id: string;
    email: string;
  };
  members: Array<{
    id: string;
    email: string;
  }>;
  createdAt: string;
  updatedAt: string;
}

export interface CreateWorkspaceRequest {
  name: string;
  description: string;
  ownerId: string;
}

@Injectable({
  providedIn: 'root'
})
export class WorkspaceService {
  private apiUrl = '/api/workspaces';

  constructor(private http: HttpClient) {}

  getAllWorkspaces(): Observable<Workspace[]> {
    return this.http.get<Workspace[]>(this.apiUrl);
  }

  getWorkspaceById(id: string): Observable<Workspace> {
    return this.http.get<Workspace>(`${this.apiUrl}/${id}`);
  }

  getWorkspacesByOwner(ownerId: string): Observable<Workspace[]> {
    return this.http.get<Workspace[]>(`${this.apiUrl}/owner/${ownerId}`);
  }

  getWorkspacesByMember(userId: string): Observable<Workspace[]> {
    return this.http.get<Workspace[]>(`${this.apiUrl}/member/${userId}`);
  }

  createWorkspace(data: CreateWorkspaceRequest): Observable<Workspace> {
    return this.http.post<Workspace>(this.apiUrl, data);
  }

  updateWorkspaceName(id: string, name: string): Observable<Workspace> {
    return this.http.patch<Workspace>(`${this.apiUrl}/${id}/name`, { name });
  }

  updateWorkspaceDescription(id: string, description: string): Observable<Workspace> {
    return this.http.patch<Workspace>(`${this.apiUrl}/${id}/description`, { description });
  }

  addMember(workspaceId: string, userId: string): Observable<Workspace> {
    return this.http.post<Workspace>(`${this.apiUrl}/${workspaceId}/members/${userId}`, {});
  }

  removeMember(workspaceId: string, userId: string): Observable<Workspace> {
    return this.http.delete<Workspace>(`${this.apiUrl}/${workspaceId}/members/${userId}`);
  }

  deleteWorkspace(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}